### PR TITLE
Fixes #17417: normalize hint modes and hint disjunctive paths

### LINKED ISSUE #17417: Hint Cut should deduplicate
#### Description of the problem

If I import a module with a Hint Cut multiple times, I get a large disjunction in Hint Cut.  It would be nice if this was deduplicated at top level, at least

#### Coq Version

8.16

### PR DESCRIPTION
**Pull request by herbelin**
We add normalization of hint paths `(p|p)` into `p` as well as non-redundancy of hint modes.

We reorganize also the normalization of paths in passing.

Fixes / closes #17417

- [x] Added / updated **test-suite**.
- [x] Added **changelog**.

### CODE REVIEW

**Comment by herbelin** [comment]
> We should use an automaton-based one instead of the native one we have today.

I don't feel competent for that, but it is unrelated to the PR afaics, right? Should we otherwise ask @mattam82? And is there a description of what this automaton would look like?

File: tactics/hints.ml [hunk 1]
```diff
@@ -442,17 +448,22 @@ let rec path_derivate hp hint =
...
   | PathSeq (p, q) ->
-    let p', q' = normalize_path p, normalize_path q in
-      if hints_path_eq p p' && hints_path_eq q q' then h
-      else normalize_path (PathSeq (p', q'))
+    (match normalize_path p with
+     | PathEmpty -> PathEmpty
+     | PathEpsilon -> normalize_path q
+     | p' ->
+     match normalize_path q with
+     | PathEmpty -> p'
```
Comment (ppedrot) [inline tactics/hints.ml:464]: This should read `PathEmpty`, not `p'`. Indeed, ` p · 0 ≡ 0`  in regexp semantics.

Comment (herbelin) [reply tactics/hints.ml:464]: Oh, right! Thanks a lot!

**Comment by ppedrot** [comment]
@coqbot run full ci

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**